securedrop_client/queue.py

```diff
@@ -169,54 +163,74 @@ def __init__(self, api_client: API, session_maker: scoped_session) -> None:
         self.main_thread.started.connect(self.main_queue.process)
         self.download_file_thread.started.connect(self.download_file_queue.process)
 
-        self.main_queue.paused.connect(self.on_queue_paused)
-        self.download_file_queue.paused.connect(self.on_queue_paused)
+        self.main_queue.paused.connect(self.on_main_queue_paused)
+        self.download_file_queue.paused.connect(self.on_file_download_queue_paused)
 
-    def logout(self) -> None:
-        if self.main_thread.isRunning():
-            logger.debug('Stopping main queue thread')
-            self.main_thread.quit()
-
-        if self.download_file_thread.isRunning():
-            logger.debug('Stopping download queue thread')
-            self.download_file_thread.quit()
-
-    def login(self, api_client: API) -> None:
-        logger.debug('Passing API token to queues')
+    def start(self, api_client: API) -> None:
+        '''
+        Start the queues whenever a new api token is provided.
+        '''
         self.main_queue.api_client = api_client
         self.download_file_queue.api_client = api_client
-        self.start_queues()
 
-    def start_queues(self) -> None:
         if not self.main_thread.isRunning():
-            logger.debug('Starting main thread')
             self.main_thread.start()
+            logger.debug('Started main queue')
 
         if not self.download_file_thread.isRunning():
-            logger.debug('Starting download thread')
             self.download_file_thread.start()
+            logger.debug('Started file download queue')
 
-    def on_queue_paused(self) -> None:
+    def stop(self) -> None:
+        '''
+        Stop the queues.
+        '''
+        if self.main_thread.isRunning():
+            self.main_thread.quit()
+            logger.debug('Stopped main queue')
+
+        if self.download_file_thread.isRunning():
+            self.download_file_thread.quit()
+            logger.debug('Stopped file download queue')
+
+    @pyqtSlot()
+    def on_main_queue_paused(self) -> None:
+        '''
+        Emit the paused signal if the main queue has been paused.
+        '''
+        logger.debug('Paused main queue')
+        self.paused.emit()
```

Comment:
this is a little hard to follow because both `ApiJobQueue` and `RunnableQueue` have signals called `paused` on them which do slightly different things: the first (signal defined on `ApiJobQueue`) signals the Controller object "hey i paused the queues, update the GUI", the other (signal defined on `RunnableQueue`) is from a queue to signal `ApiJobQueue` "hey i hit a consistent request timeout, let's stop the queues and let the GUI know". 

However it looks like these queues are pausing independently: one can be stopped while the other is running. Meanwhile the GUI shows that the server is unreachable. Is this intentional? Or is the idea that if one queue is stopped due to network failure the other will eventually stop too? 

lmk if you think I'm missing something here

Reply:
so currently, this is also how everything works on `master`, because this is a refactor pr

the way it has been implemented is that there is a `paused` signal in  `RunnableQueue`, which maintains the priority queues, and will pause when it encounters a request timeout error. meanwhile, other queues do not know about this, and cannot know unless `ApiJobQueue`, which really should be renamed to `ApiJobQueueManager`, tells them that another queue was paused. To keep things simple, the queue manager just sends a signal back to the GUI so the user can retry. It could also tell the other queue to pause by enqueuing a PauseQueueJob. I think this is risky because the other queue might at the same time be adding a PauseQueueJob if it also encounters an request timeout error.

i think changing the behavior of how this works should be discussed further and done outside of this pr, since this is a refactor only.

Reply:
one thing we could do as a small additional refactor in this PR is replace using a signal from `RunnableQueue` to using a callback that `ApiJobQueue` passes to it, so this:

```diff
--- a/securedrop_client/queue.py
+++ b/securedrop_client/queue.py
@@ -53,16 +53,15 @@ class RunnableQueue(QObject):
         ReplyDownloadJob: 17,
     }
 
-    # Signal that is emitted when processing stops
-    paused = pyqtSignal()
-
     # Signal that is emitted to resume processing jobs
     resume = pyqtSignal()
 
-    def __init__(self, api_client: API, session_maker: scoped_session) -> None:
+    def __init__(self, api_client: API, session_maker: scoped_session, on_pause_callback) -> None:
         super().__init__()
         self.api_client = api_client
         self.session_maker = session_maker
+        self.on_pause_callback = on_pause_callback
+
         self.queue = PriorityQueue()  # type: PriorityQueue[Tuple[int, ApiJob]]
         # `order_number` ensures jobs with equal priority are retrived in FIFO order. This is needed
         # because PriorityQueue is implemented using heapq which does not have sort stability. For
@@ -113,7 +112,7 @@ class RunnableQueue(QObject):
             priority, job = self.queue.get(block=True)
 
             if isinstance(job, PauseQueueJob):
-                self.paused.emit()
+                self.on_pause_callback()
                 return
 
             try:
@@ -154,8 +153,9 @@ class ApiJobQueue(QObject):
         self.main_thread = QThread()
         self.download_file_thread = QThread()
 
-        self.main_queue = RunnableQueue(api_client, session_maker)
-        self.download_file_queue = RunnableQueue(api_client, session_maker)
+        self.main_queue = RunnableQueue(api_client, session_maker, self.on_main_queue_paused)
+        self.download_file_queue = RunnableQueue(
+            api_client, session_maker, self.on_file_download_queue_paused)
 
         self.main_queue.moveToThread(self.main_thread)
         self.download_file_queue.moveToThread(self.download_file_thread)
```